src/_gettsim/policy_environment.py

```diff
@@ -1,59 +1,166 @@
+
+            # Align parameters for piecewise polynomial functions
+            params[group] = _parse_piecewise_parameters(params_one_group)
+
+        # extend dictionary with date-specific values which do not need an own function
+        params = _parse_kinderzuschl_max(date, params)
+        params = _parse_einführungsfaktor_vorsorgeaufw_alter_ab_2005(date, params)
+        params = _parse_vorsorgepauschale_rentenv_anteil(date, params)
+        functions = load_functions_for_date(date)
+
+        return PolicyEnvironment(functions, params)
+
+    def __init__(
+        self,
+        functions: list[PolicyFunction | Callable],
+        params: dict[str, Any] | None = None,
+    ):
+        self._functions = {}
+        for function in functions:
+            f = (
+                function
+                if isinstance(function, PolicyFunction)
+                else PolicyFunction(function)
+            )
+            self._functions[f.name_in_dag] = f
+
+        self._params = params if params is not None else {}
+
+    @property
+    def functions(self) -> dict[str, PolicyFunction]:
+        """The functions of the policy environment."""
+        return self._functions
+
+    @property
+    def params(self) -> dict[str, Any]:
+        """The parameters of the policy environment."""
+        return self._params
+
+    def upsert_functions(
+        self, *functions: PolicyFunction | Callable
+    ) -> PolicyEnvironment:
+        """
+        Adds to or overwrites functions of the policy environment. Note that this
+        method does not modify the current policy environment but returns a new one.
+
+        Parameters
+        ----------
+        functions:
+            The functions to add or overwrite.
+
+        Returns
+        -------
+        new_environment:
+            The policy environment with the new functions.
+        """
+        new_functions = {**self._functions}
+        for function in functions:
+            f = (
+                function
+                if isinstance(function, PolicyFunction)
+                else PolicyFunction(function)
+            )
+            new_functions[f.name_in_dag] = f
 
-    Returns
-    -------
-    params : dict
-        A dictionary with parameters from the policy environment. For more
-        information see the documentation of the :ref:`params_files`.
-    functions : dict
-        Dictionary mapping column names to functions creating the respective
-        data.
+        result = object.__new__(PolicyEnvironment)
+        result._functions = new_functions  # noqa: SLF001
+        result._params = self._params  # noqa: SLF001
 
-    """
-    # Check policy date for correct format and transfer to datetime.date
-    date = _parse_date(date)
+        return result
 
-    params = {}
-    for group in INTERNAL_PARAMS_GROUPS:
-        params_one_group = _load_parameter_group_from_yaml(date, group)
+    def replace_all_parameters(self, params: dict[str, Any]):
```

Comment:
Just a mental note and link to #781 that we would probably want something like `upsert_params` to just change tiny portions. But precise interface will depend on how parsing of params and their inclusion in DAG will look like.